src/main/java/org/testng/TestNG.java

```diff
@@ -2070,14 +2072,16 @@ public void setServiceLoaderClassLoader(URLClassLoader ucl) {
    * Used to test ServiceClassLoader
    */
   private void addServiceLoaderListener(ITestNGListener l) {
-    m_serviceLoaderListeners.add(l);
+    if (! m_serviceLoaderListeners.containsKey(l.getClass())) {
```

Comment:
Thinking out loud: now that we have all these listener maps scattered a bit everywhere, we might want to encapsulate this in a `ListenerContainer`. Have just one field of this type on this class and `m_listenerContainer.addServiceLoaderListener(...)`.

Don't do this in this PR, though (and not sure it's worth doing, just wondering).